src/goto-instrument/contracts/dynamic-frames/dfcc_swap_and_wrap.cpp

```diff
@@ -0,0 +1,313 @@
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmarsd@amazon.com
+
+\*******************************************************************/
+
+#include "dfcc_swap_and_wrap.h"
+
+#include <util/config.h>
+#include <util/expr_util.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
+#include <util/fresh_symbol.h>
+#include <util/mathematical_expr.h>
+#include <util/mathematical_types.h>
+#include <util/namespace.h>
+#include <util/optional.h>
+#include <util/pointer_expr.h>
+#include <util/pointer_offset_size.h>
+#include <util/pointer_predicates.h>
+#include <util/std_expr.h>
+
+#include <goto-programs/goto_functions.h>
+#include <goto-programs/goto_inline.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/instrument_preconditions.h>
+#include <goto-programs/link_to_library.h>
+#include <goto-programs/remove_skip.h>
+
+#include <ansi-c/c_expr.h>
+#include <ansi-c/cprover_library.h>
+#include <goto-instrument/contracts/cfg_info.h>
+#include <goto-instrument/contracts/instrument_spec_assigns.h>
+#include <goto-instrument/contracts/utils.h>
+#include <linking/static_lifetime_init.h>
+
+dfcc_swap_and_wrapt::dfcc_swap_and_wrapt(
+  goto_modelt &goto_model,
+  messaget &log,
+  dfcc_utilst &utils,
+  dfcc_libraryt &library,
+  dfcc_instrumentt &instrument,
+  dfcc_spec_functionst &spec_functions,
+  dfcc_contract_handlert &contract_handler)
+  : goto_model(goto_model),
+    log(log),
+    message_handler(log.get_message_handler()),
+    utils(utils),
+    library(library),
+    instrument(instrument),
+    spec_functions(spec_functions),
+    contract_handler(contract_handler),
+    ns(goto_model.symbol_table)
+{
+}
+
+// static map
+std::map<irep_idt, std::pair<irep_idt, dfcc_contract_modet>>
+  dfcc_swap_and_wrapt::cache;
+
+void dfcc_swap_and_wrapt::swap_and_wrap(
+  const dfcc_contract_modet contract_mode,
+  const irep_idt &function_id,
+  const irep_idt &contract_id,
+  std::set<irep_idt> &function_pointer_contracts,
+  bool allow_recursive_calls)
+{
+  log.debug() << "->swap_and_wrap(" << function_id << "," << contract_id << ")"
+              << messaget::eom;
+  auto found = cache.find(function_id);
+  if(found != cache.end())
+  {
+    auto &pair = found->second;
+
+    // same swap already performed
+    if(pair.first == contract_id && pair.second == contract_mode)
+      return;
+
+    // already swapped with something else, abort
+    log.error() << " multiple attempts to swap and wrap function '"
+                << function_id << "':" << messaget::eom;
+    auto mode1 =
+      (pair.second == dfcc_contract_modet::REPLACE) ? "REPLACE" : "CHECK";
+    log.error() << "with '" << pair.first << "' in mode " << mode1
+                << messaget::eom;
+    auto mode2 =
+      (contract_mode == dfcc_contract_modet::REPLACE) ? "REPLACE" : "CHECK)";
+    log.error() << "with '" << contract_id << "' in mode " << mode2
+                << messaget::eom;
+    throw 0;
```

Comment:
Is this an internal error (should be an `INVARIANT`) or something to be triggered by user code (should be `invalid_input_exceptiont`)

Reply:
this could be a user error